README.md
# 「回声工坊辅助插件」的帮助文档

这是文字视频（例如TRPG replay、视觉小说）制作软件「回声工坊」的辅助扩展，围绕整理剧本文件（跑团Log文件）提供了很多便利功能

- [软件“回声工坊”的介绍视频](https://www.bilibili.com/video/BV1GY4y1H7wK/)
- [软件“回声工坊”的Github仓库](https://github.com/DanDDXuanX/TRPG-Replay-Generator)
- [本扩展的简单介绍视频](https://www.bilibili.com/video/BV1Xv4y1u7RC/)
- [本扩展的图文帮助文档](https://www.wolai.com/oPvjzFX3gQDZjV5JLAxpuR)
- [本扩展的最新开发进展](https://www.wolai.com/544d41yz3DEFyqSy2YquHW)

## 基础功能使用方法

1. 去官网安装[vscode](https://code.visualstudio.com/)（不过你大概率是在vscode里面看到这句话，所以是一句废话）。**注意，这玩意儿是免费的，不要下载到收费的盗版软件。**
2. **将要高亮的剧本文件的扩展名改为.rgl，txt文件是不会高亮的。**
3. 安装该扩展之后，在弹窗中使用该扩展提供的“RGL Dark Theme”或者“RGL Light Theme”主题。不用自带主题的话可能高亮起来怪怪的。当然，可以切换其他主题，同样有语法高亮。
4. 开始润色你的剧本文件吧！不用考虑“那么多功能全部看完好麻烦”，你需要用到的时候再去看也行。

Tips：
- rgl文件的右键菜单集成了一些常用命令

## 高级功能使用方法

1. 按照下文的方法进行本扩展的设置，一般来说，“文件路径”那一栏设置最好都给设置好
2. 按照下文列出的功能去试试吧

### 全局配置与局部配置

点击左下角的齿轮图标或者使用 `ctrl+,`快捷键可以打开设置面板，在搜索框中搜索本插件名称（或者在旁边的“Extensions”一栏中找这一项），里面有着相关配置。

以下配置可以写在rgl文件中作为局部配置，存在多条配置时，会使用第一个。

- 若不存在局部配置，则使用全局配置。
- `#md`和 `#MediaDefinition`两种写法都可以，且**不区分大小写**。
- 优先使用当前文件作为log文件，若未打开文件，则使用全局配置的log文件路径。
- 支持相对路径，局部配置相对于当前log文件，全局配置相对于当前log文件或者全局配置中的log文件（优先取当前log文件的路径）
- 代码片段 `current directory`:快速输入当前文档所在目录，便于进行局部配置
- 代码片段 `quickly configurate`:快速在本文件中配置与本文件同目录下的媒体定义文件media.txt和角色配置表characters.tsv

```
#! RplGenCore.exe
#md 媒体定义文件路径
#MediaDefinition 媒体定义文件路径

#ct 角色配置表路径
#CharacterTable 角色配置表路径

#op 输出路径
#Output 输出路径

#tl 导出XML时的时间轴文件的路径
#TimeLine 导出XML时的时间轴文件的路径

#ll 悬停提示和代码诊断的每行限制字数
#LineLength 悬停提示的每行限制字数

#tl 代码诊断所使用的对话行内容总字数
#TotalLength 代码诊断所使用的对话行内容总字数
```

## 本扩展提供的功能

### 语法高亮与颜色主题

- `RGL Dark Theme`：暗色主题
- `RGL Light Theme`：亮色主题

### 支持的代码片段

| 代码片段                | 说明                                                                                                                                                                                                                                                                                                                                         |
| ----------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| 内建动画行              | 支持 `<dice>`和 `<hitpoint>`的快速输入。例如：输入dice4可以快速输入含有4组参数的骰子行，输入hp可以快速输入生命值行                                                                                                                                                                                                                       |
| 设置行                  | 输入set可以在弹出的列表中选择需要输入的设置                                                                                                                                                                                                                                                                                                  |
| 背景行                  | 输入background + 切换方式的名称或者bg+切换方式开头字母就可以快速输入背景行。例如 `bgr`或 `background replace`为 `<background><replace=0>:背景媒体`                                                                                                                                                                                     |
| 音效框                  | 输入audio可以快速输入音效框，即 `{音效;*时间}`                                                                                                                                                                                                                                                                                             |
| 正则表达式片段          | 便于复制使用的常用正则表达式片段，触发词前缀是regex。分别是“将跑团记录着色器得到的尖括号替换为方括号与冒号”，“批量给对话行末尾添加待处理星标“{*}”。”，“给某个特定角色对话行末尾加特定音效。”（现在推荐直接使用对应的命令 `TRGL: Replace Angle Brackets`、`TRGL: Add Asterisk Marks`、`TRGL: Add Sound Effects In Batches`、 ） |
| `fold mark`           | 快速输入折叠标记                                                                                                                                                                                                                                                                                                                             |
| `current directory`   | 快速输入当前文档所在目录，便于进行局部配置                                                                                                                                                                                                                                                                                                   |
| `quickly configurate` | 快速在本文件中配置与本文件同目录下的媒体定义文件media.txt和角色配置表characters.tsv                                                                                                                                                                                                                                                          |
| `animation`           | 常驻立绘行                                                                                                                                                                                                                                                                                                                                   |
| `clear`               | 聊天窗内容清除行                                                                                                                                                                                                                                                                                                                             |
| `set FreePos`         | 设置自由位置                                                                                                                                                                                                                                                                                                                                 |

### 折叠标记

- 在想要折叠的起始行和末尾行分别加上 `#s`和 `#e`的注释，即可对中间的内容进行折叠
- 可以嵌套，而且这两个边界标记后面可以加注释，例如 `#s 这是折叠的起始行，意思是start`，`#e 这是折叠的末尾行，意思是end`。
- **媒体定义文件**支持按照媒体类型折叠以及折叠标记折叠（#s开头和#e结尾，使用方式同上）

### 悬停提示

鼠标移动到rgl文件的各个元素并悬停时，会出现相应的说明文字。可以在设置中关闭。

可以使用 `#ll 50`或者 `#lineLength 50`将单行字数限制设定为50，这里的数字可以改成其他整数，设置为负数将被视作正无穷

### 自动补全

- **角色差分**：你在角色名后输入“.”即可自动补全角色差分，只要你正确配置了角色配置表（支持tsv和xlsx格式）
- **角色名**：输入对话框的名字框时触发，多个角色时也支持
- **背景行**：在 `<background>`后输入英文冒号触发
- **音效**：输入左花括号 `{`触发
- **立绘媒体**：常驻立绘行(`<animation>`)自动补全立绘
- **气泡媒体**：常驻气泡行(`<bubble>`)自动补全气泡
- **聊天窗媒体**：聊天窗内容清除行(`<clear>`)自动补全聊天窗
  
### 快捷换行

使用Ctrl+Enter快捷键插入和当前行同角色名的对话行，即换行时保留当前行角色框

例如：

```
[角色A,角色B]:发言内容1。发言内容2。
```

在两段发言内容中间用Ctrl+Enter快捷键进行换行，会变成下面这样：

```
[角色A,角色B]:发言内容1。
[角色A,角色B]:发言内容2。
```


### 代码诊断及快速修复

- 当角色或者差分未在角色配置表中定义的时候，报错
- 当背景行使用的背景未在媒体定义文件中定义时，报错
- 当对话行单行字数或者总字数超出用户设置的上限时，警告（可将对应字数设置为负数以关闭此功能）
- 自动检测多音字并标注，以防止语音合成不合预期，可在设置中调整多音字表和提示等级，并提供“在行尾添加合成指定文本的语音的标志便于修改”的快速修复

代码诊断新增音效框相关：

- 检查语音合成指定文本只能包含`，。：？！“”`等中文符号
- 检查音效框的文件路径是否正确
- 检查音效框使用的是否为wav文件
- 检查音效媒体是否在媒体定义文件中已经定义

### 状态栏

可以在状态栏里看到预估视频时长等统计信息，点击状态栏可以使用

### 侧边栏

点击侧边栏回声工坊图标即可打开回声工坊资源管理器

|视图|说明|
|---|---|
|大纲|会将背景行视作跳转节点。侧边栏大纲可以使用折叠标记作为大纲节点，可嵌套（折叠标记正确闭合才会得到正确大纲），在【设置】->【侧边栏】中可以找到相应设置，默认使用背景行作为大纲节点|
|角色管理|可以通过侧边栏查看当前rgl文件有哪些角色差分（带*号）以及角色配置表里面有哪些角色差分，点击按钮手动刷新，可插入到剧本文件中|
|背景|会读取已配置的媒体定义文件得到可用背景媒体，点击可插入到剧本文件中|
|音效|会读取已配置的媒体定义文件得到可用音效媒体，点击可插入到剧本文件中|

### 命令

命令的使用方法是：ctrl+shift+p唤出命令面板，输入命令名字，enter键确认

打开命令面板的其他方式：

- 右键菜单中“命令面板”选项
- 菜单栏【帮助】>【显示所有命令】

#### `TRGL: Open Document`

可以快速打开配置的回声工坊软件路径下的 `README.md`文件，方便查看回声工坊的文档

#### `TRGL: Chat With DiceBot`

可以询问答疑骰郎伊可（Exception）问题

#### `TRGL: Correct Typos`

调用阿里云key进行文本纠错的功能，详细使用方法见：https://www.wolai.com/yxchangingself/D5svMQ4DWThCzUyauUbSb

对于已经配置好语音合成的用户来说，你需要做的就一件事：[开通文本纠错服务](https://help.aliyun.com/document_detail/179148.html)，key都是用以前的

#### `TRGL: Count`

可以统计对话行和骰子行的行数，并将rgl文件中提到的所有角色和背景不重复地列出，方便填写角色配置表和定义背景媒体，同时可以估算生成的视频的时间。

只统计选中部分的内容（行号为相对行号），当未选中任何文本时，统计整个文件

#### `TRGL: Check Dialog Line Length`

找出超出指定字数的对话行

#### `TRGL: Replace Angle Brackets`

将[QQ跑团记录着色器](https://logpainter.kokona.tech/)得到的尖括号替换为方括号与冒号

#### `TRGL: Add Asterisk Marks`

**（已经弃用，请使用`TRGL: Edit Audio Box`命令代替）**

批量给对话行末尾添加待处理星标“{*}”，拥有多个选项


CHANGELOG.md
# 更新日志

此项目的所有显著更改都将记录在此文件中。

格式基于[维护更新日志（Keep a Changelog）](https://keepachangelog.com/en/1.0.0/)，
并且该项目遵循 [语义版本控制（Semantic Versioning）](https://semver.org/spec/v2.0.0.html)。

## [1.25.2] - 2023-06-16

- 修复每次开启vscode时会多出一个「回声工坊」终端的bug

## [1.25.1] - 2023-03-17

- 给进度条加了个更加直观的显示方式：方块进度条

## [1.25.0] - 2023-03-17

- 在悬停提示中添加了进度条预览

## [1.24.1] - 2023-03-17

- 现在插入可用角色时，直接替换选中文字，而非插入到前方
- 新增了插入无角色框文本的按钮

## [1.24.0] - 2023-03-07

现在可以使用Ctrl+Enter快捷键插入和当前行同角色名的对话行，即换行时保留当前行角色框

例如：

```
[角色A,角色B]:发言内容1。发言内容2。
```

在两段发言内容中间用Ctrl+Enter快捷键进行换行，会变成下面这样：

```
[角色A,角色B]:发言内容1。
[角色A,角色B]:发言内容2。
```

## [1.23.1] - 2023-02-23

- 修复侧边栏角色管理器读取tsv角色配置表时未删除表头以及将空白行视作一个角色的bug

## [1.23.0] - 2023-02-02

- 侧边栏可用背景视图支持悬停预览
- 背景行支持悬停预览图片

## [1.22.0] - 2023-01-12

`TRGL: Edit Content`命令新增选项：

- 省略号规范化：将连续出现的两个以上的中英文逗号和句号转化为中文省略号
- 空格转化为逗号：将空格转化为中文逗号

## [1.21.7] - 2023-01-08

- QQ聊天记录转化多适配了一个日期之间以连字符分割的格式（为什么QQ复制导出会有这么多不同格式啊！）

## [1.21.6] - 2022-12-23

- QQ聊天记录转化多适配了一个带上下午的格式

## [1.21.5] - 2022-12-21

- 图标优化
- 修复侧边栏悬停提示一直加载的bug

## [1.21.4] - 2022-12-17

- 右键菜单中的回声工坊命令，现在除了可以在rgl文件中使用外，也可以在txt文件中使用了。当然无论是这个版本之前还是之后，都可以用命令面板调用命令

## [1.21.3] - 2022-12-12

- 局部配置和全局配置的文件路径现在支持相对路径了，局部配置相对于当前log文件，全局配置相对于当前log文件或者全局配置中的log文件（优先取当前log文件的路径）
- 整理右键菜单命令，将本扩展的命令分类整理，方便查找使用

## [1.21.2] - 2022-12-10

- `TRGL: Edit Content`命令可以只编辑已选中的行，若未选中任何行，则编辑整个文件
- `TRGL: Edit Audio Box`命令同上

## [1.21.1] - 2022-12-10

- `TRGL: Edit Content`命令新增选项：对对话行的内容进行自定义正则替换。仅作用于对话行的内容部分，所以正则表达式无需写匹配角色框的部分

## [1.21.0] - 2022-12-09

- 修改右键菜单内容
- 删除命令`TRGL: Add Asterisk Marks`，请使用`TRGL: Edit Audio Box`替代
- 删除命令`TRGL: Add Sound Effects In Batches`，请使用`TRGL: Edit Audio Box`替代
- 新增命令`TRGL: Edit Content`，批量修改对话行内容

#### `TRGL: Edit Content`

批量编辑对话行内容

- 注释掉场外交流行
- 注释掉指令行
- 给场外交流行的主角色添加「场外」差分
- 删除注释掉的对话行
- 左右引号顺序修正
- 单双引号修正
- 替换为直角引号
- 为RP内容分段

## [1.20.3] - 2022-12-08

为导出XML和合成语音两个命令添加了输出文件的二次确认，二次确认中修改的路径不会修改你的设置，只是个临时修改

## [1.20.2] - 2022-12-08

命令`TRGL: Edit Audio Box # 批量编辑音效框`新增两个选项：

- 给指定角色删除文件路径音效
- 给指定角色删除所有音效

## [1.20.1] - 2022-12-08

把插件最低兼容的vscode版本升级到了1.74.0，可能会用上vscode更新的一些新功能。vscode应该会自动更新，如果没有更新，也可以在顶部菜单【帮助】一栏里面找到【检查更新】，对具体更新内容感兴趣可以看[官方更新日志](https://code.visualstudio.com/updates/v1_74)

## [1.20.0] - 2022-12-07

新增命令`TRGL: Edit Audio Box # 批量编辑音效框`，用于上位替代`TRGL: Add Sound Effects In Batches # 批量添加音效`和`TRGL: Add Asterisk Marks # 批量添加语音待处理星标{*}`，后二者即将弃用

新命令不再是简单的正则替换，而是可以真正指定特定角色（无论是否有差分，当前角色框是否有其他角色都可以）

同时，新增删除指定角色的指定音效的选项，音效框也不会误删除

#### `TRGL: Edit Audio Box`

批量编辑音效框，拥有多个选项：

- 给全部对话行添加{*}
- 只给没有音效框的对话行行添加{*}
- 给指定角色添加指定音效(默认为{*})
- 给指定角色删除指定音效(默认为{*})
- 去掉纯标点符号行的{*}

## [1.19.3] - 2022-12-06

- 修复NA不被认作媒体对象的问题。现在`{NA}`和`{NA;*时间}`的写法不会报错了

## [1.19.2] - 2022-12-06

- `TRGL: Migrate Log`命令已弃用，请使用`TRGL: Convert Log`命令进行活字引擎和回声工坊Log格式互相转换

## [1.19.1] - 2022-12-06

- `TRGL: Convert Log`新增对QQ聊天消息（直接复制和导出两种都行）和已经染色的跑团记录（塔、赵、溯洄）的转换

## [1.19.0] - 2022-12-05

新增命令：

#### `TRGL: Convert Log`

转换跑团Log格式，对当前文件使用，在旁边生成一个转换好的回声工坊log文件，目前支持：

- 猫爷TRPG => 回声工坊
- 菠萝文字团平台txt => 回声工坊
- 菠萝文字团平台json => 回声工坊

## [1.18.7] - 2022-12-04

代码诊断新增音效框相关：

- 检查语音合成指定文本只能包含`，。：？！“”`等中文符号
- 检查音效框的文件路径是否正确
- 检查音效框使用的是否为wav文件
- 检查音效媒体是否在媒体定义文件中已经定义

## [1.18.6] - 2022-12-04

- 前一个更新没考虑到删除待合成星标的情况，加上了对应选项“删除某个角色后的所有音效框（会误伤普通音效，待优化）”

## [1.18.5] - 2022-12-04

- `TRGL: Add Asterisk Marks`命令新增选项：把某个角色已经合成的语音框删除。可以先添加全部{*}，再删除某个特定角色的语音框，实现只给某个角色不添加星标


## [1.18.4] - 2022-12-04

关于`TRGL: Copy Log`命令的一些小修复：
- 修复了复制内容中会出现角色差分以及手动换行符号的问题
- 可以只复制选中的文本，如果没有选中，则复制当前打开的全文
- 支持快捷键“Ctrl+Shift+C”

## [1.18.3] - 2022-12-04

- 新增“TRGL: Copy Log # 以不同格式复制Log文件文本”命令，支持复制为“软件「朗读女」的「多角色朗读」的格式：[角色名]说话内容”以及“无角色框的内容文本”

## [1.18.2] - 2022-12-02

- 修复了文本纠错功能无法识别文本中的星号（*）的问题（为什么会报错呢？奇怪）

## [1.18.1] - 2022-12-01

- 代码片段新增回声工坊1.1.0版的`inline_method_apply`设置行
- 骰子行新增对出目大于骰面数的代码诊断

## [1.18.0] - 2022-11-30

- 新增调用阿里云key进行文本纠错的功能，详细使用方法见：https://www.wolai.com/yxchangingself/D5svMQ4DWThCzUyauUbSb

## [1.17.1] - 2022-11-30

- 修复了多音字诊断的下划线标注错位的问题

## [1.17.0] - 2022-11-29

- 新增多音字诊断和快速修复，可在设置中调整多音字表和提示等级，并提供“在行尾添加合成指定文本的语音的标志便于修改”的快速修复

## [1.16.8] - 2022-11-29

- 修复设置行行数统计错误的问题（刚刚那次没有完全修复）

## [1.16.7] - 2022-11-29

- 修复设置行行数统计错误的问题


src/convertLog.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

//命令
export function convertLog(){

    //获取文本
    let editor = vscode.window.activeTextEditor;
    if(!editor){
        vscode.window.showInformationMessage("请打开你要转换的文件");
        return;
    }
    
    let doc = editor.document;
    let text = doc.getText();


    //转换Log
    let optMaoYe = "猫爷TRPG => 回声工坊";
    let optBoLuoTxt = "菠萝txt => 回声工坊";
    let optBoLuoJson = "菠萝json => 回声工坊";
    let optQQ = "QQ聊天记录 => 回声工坊";
    let optPainted = "已染色记录 => 回声工坊";
    let optPaintedTaDice = "塔骰已染色记录 => 回声工坊";
    let optEngine2Generater = "活字引擎 => 回声工坊";
    let optGenerater2Engine = "回声工坊 => 活字引擎";
    vscode.window.showQuickPick([{
        label: optMaoYe,
        description: '猫爷TRPG：https://maoyetrpg.com/',
        detail: '并不是所有骰子都进行了转换；某些换行需要手动处理',
    }, {
        label: optBoLuoTxt,
        description: '菠萝：https://boluo.chat/',
        detail: '菠萝可导出多种格式，此选项处理的是txt格式。由于是行内骰子，不易解析，所以没有处理。建议导出到json格式来处理'
    }, {
        label: optBoLuoJson,
        description: '菠萝：https://boluo.chat/',
        detail: '菠萝可导出多种格式，此选项处理的是json格式。'
    }, {
        label: optQQ,
        description: 'QQ未处理的跑团记录',
        detail: '支持直接从聊天窗口复制、消息管理器导出的记录(还是不要两者混用了，混用有点bug)'
    }, {
        label: optPainted,
        description: '已经用染色器网站处理过的跑团记录',
        detail: '支持赵系(https://logpainter.trpgbot.com/)、溯洄系(https://logpainter.kokona.tech)染色器网站'
    }, {
        label: optPaintedTaDice,
        description: '已经用染色器网站处理过的跑团记录，塔骰专属',
        detail: '支持塔骰染色记录'
    }, {
        label: optEngine2Generater,
        description: '活字引擎3',
        detail: ''
    }, {
        label: optGenerater2Engine,
        description: '',
        detail: ''
    }],{
        title:"选择log转换模式",
        placeHolder:"将会在原文件旁边生成转换好的rgl文件"
    }).then(value => {
        if(!value){
            return;
        }
        switch(value?.label){
            case optMaoYe:
                text = convertLogFromMaoYe(text);
                break;
            case optBoLuoTxt:
                text = convertLogFromBoLuoTxt(text);
                break;
            case optBoLuoJson:
                text = convertLogFromBoLuoJson(text);
                break;
            case optQQ:
                text = convertLogFromQQ(text);
                break;
            case optPainted:
                text = convertLogFromPainted(text);
                break;
            case optPaintedTaDice:
                text = convertLogFromPaintedTaDice(text);
                break;
            case optEngine2Generater:
                text = convertEngine2Generater(text);
                break;
            case optGenerater2Engine:
                text = convertGenerater2Engine(text);
                break;
        }
        //写入文件
        let newFileNamePath = path.join(path.dirname(doc.fileName),`${path.basename(doc.fileName,path.extname(doc.fileName))}_${new Date().getTime()}.rgl`);

        fs.writeFileSync(newFileNamePath,text);
        vscode.commands.executeCommand("vscode.open",vscode.Uri.file(newFileNamePath));
    });
}

//猫爷TRPG
export function convertLogFromMaoYe(text:string){
    return text
    .replaceAll(/(.*?) \d\d\d\d\/\d\d\/\d\d \d\d:\d\d:\d\d[\n\r]+    (.*)/gm,"[$1]:$2")//注意回车换行符（\r\n）
    // .replaceAll(new RegExp("(.*?) \\d\\d\\d\\d\\/\\d\\d\\/\\d\\d \\d\\d:\\d\\d:\\d\\d[\\n\\r]+    (.+)","gm"),"[$1]:$2")//注意回车换行符（\r\n）
    .replaceAll(/#/g,"")
    .replaceAll(/<br>/g,"")//因为猫爷TRPG的Log常常在奇怪的地方加<br>，所以直接去掉
    .replaceAll(/\(/g,"（")//英文括号转中文括号
    .replaceAll(/\)/g,"）")//英文括号转中文括号
    .replaceAll(/^\[骰子\]:(.*)[:：].*[Dd](\d*)\s*=\s*(\d*)\/(\d*).*$/gm,"# $&\n<dice>:($1,$2,$4,$3)")//骰子行转换
    .replaceAll(/^\[骰子\]:(.*)[:：] .*\(\d+[Dd](\d+)\s*=\s*\d+\s*\)=\d+=(\d+).*$/gm,"# $&\n<dice>:($1,$2,NA,$3)")
    .replaceAll(/^.+:button756.+$/mg,"# $&")//注释掉掷骰指令行
    .replaceAll(/^\[.*\]:\..*$/mg,"# $&")//注释掉掷骰指令行
    .replaceAll(/^\[.*\]:<img src=.*>$/mg,"# $&");//注释掉图片行
}

//菠萝文字团平台txt格式
export function convertLogFromBoLuoTxt(text:string){
    return text
    .replaceAll(/\[\d\d\d\d-\d\d-\d\d \d\d:\d\d\] <(.+)>  ((.|\n)+?)(?=\[\d\d\d\d-\d\d-\d\d \d\d:\d\d\])/gm,"[$1]:$2")
    .replaceAll(/^\[\d\d\d\d-\d\d-\d\d \d\d:\d\d\]   \* (\S+) (.+)$/gm,"[$1.行动]:$2")//标记动作
    .replaceAll(/\$ \[附件\]\(.*\)/g,"\n# $&");//注释附件
}

//菠萝文字团平台json格式
export function convertLogFromBoLuoJson(text:string){
    try {
        let lines = JSON.parse(text);
        let content = "";
        for(let line of lines){
            if(line.isAction){
                content += `${line.entities.length > 1?"# ":""}[${line.name}.行动]:${line.entities[0].text.replaceAll("#","").replaceAll("\n","#")}\n`;
            }else{
                content += `${line.entities.length > 1?"# ":""}[${line.name}]:${line.text.replaceAll("#","").replaceAll("\n","#")}\n`;
            }
            if(line.entities.length > 1){
                //不止一条则说明有骰子
                let dices:string[] = [];
                for(let entity of line.entities){
                    if(entity.type === "Expr" && (entity.node.type === "Roll" || entity.node.type === "CocRoll")){
                        let node = entity.node;
                        dices.push(`(${line.text.replaceAll(/[\{\}]/g,"")},${node.face * node.counter},${node.targetValue?node.targetValue:"NA"},${node.value})`);
                    }
                }
                content += `<dice>:${dices.join(",")}\n`;
            }
        }
        return content;
    } catch (error) {
        vscode.window.showErrorMessage("解析json时出错，请确保你用的是json文件");
    }
    return text;
}

//QQ聊天记录
export function convertLogFromQQ(text:string){

    //直接复制的格式，日期和时间在昵称后面：昵称(QQ号)
    let regexCopy = /^(\S+?)\s+(\d\d\d\d[\/\-]\d\d[\/\-]\d\d? )?\d\d?:\d\d:\d\d( #\d+)?/m;
    //消息管理器导出的格式
    let regexExport = /^(\d\d\d\d[\/\-]\d\d?[\/\-]\d\d?.*\d\d?:\d\d:\d\d) (.+)(\([^\(\n]+\)|<[^\(\n]+>)/m;

    let lines = text.split('\n');
    let curPC = "";
    let newText = "";
    for(let line of lines){
        line = line.trim();
        if(line === "" || /^\[图片\]$/m.test(line)){
            continue;
        }
        let r:RegExpExecArray | null;
        if((r = regexCopy.exec(line)) && r[2]){//如果分组2没匹配到说明把时间当做PC名匹配了
            //匹配到，则说明切换角色了
            curPC = r[1].replaceAll("(","（").replaceAll(")","）").replaceAll("[","【").replaceAll("]","】");
        }else if(r = regexExport.exec(line)){
            curPC = r[2].replaceAll("(","（").replaceAll(")","）").replaceAll("[","【").replaceAll("]","】");
        }else{
            //全都未匹配上，说明是内容行
            let content = line.replaceAll("(","（").replaceAll(")","）").replaceAll("#","");
            if(content.startsWith(".") || content.startsWith("。")){
                newText += `\n# [${curPC}]:${content}`;//注释指令行
            }else{
                newText += `\n[${curPC}]:${content}`;
            }
        }
    }
    return newText;
}

//已经染色的记录
//赵，溯洄
export function convertLogFromPainted(text:string){
    //^((\d\d\d\d\/\d\d\/\d\d? )?\d\d?:\d\d:\d\d)?\s*<(.+)>\s*(.+)$
    return text
    .replaceAll(/^((\d\d\d\d[\/\-]\d\d[\/\-]\d\d? )?\d\d?:\d\d:\d\d)?\s*<(.+)>\s*(.+)$/gm,"[$3]:$4")
    .replaceAll(/#/g,"");
}

//塔骰
export function convertLogFromPaintedTaDice(text:string){

    //普通对话行
    let regexDialog = /^(<\d{4}-\d{2}-\d{2} \d+:\d+:\d+\.0>\s*)?\[(.*?)\]:\s*(.*)$/m;
    //场外行
    let regexOB = /^\(\[(.+)\]:\s*(.+)\)$/m;

    let lines = text.split('\n');
    let curPC = "";
    let newText = "";
    for(let line of lines){
        line = line.trim();
        if(line === ""){
            continue;
        }
        let r:RegExpExecArray | null;
        if(r = regexDialog.exec(line)){
            //匹配到，则说明切换角色了
            curPC = r[2].replaceAll("(","（").replaceAll(")","）").replaceAll("[","【").replaceAll("]","】");
            let content = r[3].replaceAll("(","（").replaceAll(")","）");
            if(content.startsWith("#")){
                newText += `\n[${curPC}.行动]:${content.replaceAll("#","")}`;
            }else{
                newText += `\n[${curPC}]:${content.replaceAll("#","")}`;
            }
        }else if(r = regexOB.exec(line)){
            curPC = r[1].replaceAll("(","（").replaceAll(")","）").replaceAll("[","【").replaceAll("]","】");
            let content = r[2].replaceAll("(","（").replaceAll(")","）").replaceAll("#","");
            newText += `\n[${curPC}.吐槽]:（${content}）`;
        }else{
            //未匹配上，说明是内容行
            let content = line.replaceAll("(","（").replaceAll(")","）").replaceAll("#","");
            if(line.startsWith("#")){
                //如果是动作行
                newText += `\n[${curPC}.行动]:${content}`;
            }else{
                newText += `\n[${curPC}]:${content}`;
            }
        }
    }
    return newText;
}


//活字引擎 => 回声工坊
function convertEngine2Generater(text:string){
    //转换对话行
    return text.replaceAll(/^<(.+?)(（(.+)）)?>(.+)(\n【读作】(.+))?/gm,"[$1.$3]:$4{*$6}")
    .replaceAll(/^\[(.+)\.\]:/gm,"[$1]:")

    //转换背景行
    .replaceAll(/^【背景】(.+)$/gm,"<background>:$1")
    //BGM
    .replaceAll(/^【BGM】(.+)$/gm,"<set:BGM>:$1")
    .replaceAll(/^【停止BGM】(.+)?$/gm,"<set:BGM>:stop")

    
    //骰子
    .replaceAll(/^【骰子】（(.+?)）D(\d+)=(\d+)(\/(\d+))?(；（(.+?)）D(\d+)=(\d+)(\/(\d+))?)?(；（(.+?)）D(\d+)=(\d+)(\/(\d+))?)?(；（(.+?)）D(\d+)=(\d+)(\/(\d+))?)?$/gm,"<dice>:($1,$2,$5,$3),($7,$8,$11,$9),($13,$14,$17,$15),($19,$20,$23,$21)")
    .replaceAll(/,\(,,,\)/gm,"")
    .replaceAll(/\((.+?),(\d+?),,(\d+?)\)/gm,"($1,$2,NA,$3)")

    //注释
    .replaceAll(/^\/\/(.+)$/gm,"#$1")

    //其余行
    .replaceAll(/(^【.+$)/gm,"#$1");
}

//回声工坊 => 活字引擎
function convertGenerater2Engine(text:string){

    //其余行
    return text.replaceAll(/^(<set:.+)$/gm,"//$1")
    .replaceAll(/^(<hitpoint>:.+)$/gm,"//$1")

    //转换对话行
    .replaceAll(/^\[(.+?)(\.(.+))?(,.+)?\]:(.+?)(\{.+\})?$/gm,"<$1（$3）>$5")
    .replaceAll(/^<(.+?)（）>/gm,"<$1>")

    //转换背景行
    .replaceAll(/^<background>(<.+>)?:(.+)$/gm,"【背景】$2")

    //BGM
    .replaceAll(/^<set:BGM>:(.+)$/gm,"【BGM】$1")
    .replaceAll(/^<set:BGM>:stop$/gm,"【停止BGM】")

    //骰子
    .replaceAll(/^<dice>:\((.+?),(.+?),(.+?),(.+?)\)(,\((.+?),(.+?),(.+?),(.+?)\))?(,\((.+?),(.+?),(.+?),(.+?)\))?(,\((.+?),(.+?),(.+?),(.+?)\))?$/gm,"【骰子】（$1）D$2=$4/$3；（$6）D$7=$9/$8；（$11）D$12=$14/$13；（$16）D$17=$19/$18")
    .replaceAll(/；（）D=\//gm,"")
    .replaceAll("/NA","")

    //注释
    .replaceAll(/^#(.+)$/gm,"//$1");
}